drivers/mcp2515/mcp2515_spi.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2016 OTA keys
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_mcp2515
+ * @{
+ *
+ * @file
+ * @brief       MCP2515 can spi driver
+ *
+ * @author      Toon Stegen <toon.stegen@altran.com>
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "mcp2515_spi.h"
+#include "mcp2515_defines.h"
+
+#include "periph/gpio.h"
+#include "periph/spi.h"
+
+#include "xtimer.h"
+#include "irq.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+int mcp2515_spi_init(const candev_mcp2515_t *dev)
+{
+    int res;
+
+    /* Configure SPI */
+    res = spi_init_cs(dev->conf->spi, dev->conf->cs_pin);
+    if (res != SPI_OK) {
+        DEBUG("spi_init_master: error initializing SPI_%i device (code %i)\n",
+              dev->conf->spi, res);
+        return -1;
+    }
+    return 0;
+}
+
+int mcp2515_spi_reset(const candev_mcp2515_t *dev)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, false,
+                      MCP2515_SPI_RESET);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_read(const candev_mcp2515_t *dev, uint8_t addr, uint8_t *buf,
+                     unsigned int len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_READ);
+    spi_transfer_regs(dev->conf->spi, dev->conf->cs_pin, addr, NULL,
+                      (void *)buf, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_read_rxbuf(const candev_mcp2515_t *dev, uint8_t mailbox,
+                           void *buf, uint8_t len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_READ_RXBUF | (mailbox << 2));
```

Comment:
Why the shift?

Reply:
Can't really find it in the datasheet atm (I must be overlooking it), but this is to select what mailbox to read from. There are two mailboxes available for reception.

Reply:
It is explained int TABLE 12-1:SPI INSTRUCTION SET, can you add a comment pointing out to this?